src/Microsoft.ML.CpuMath/CpuMathUtils.netcoreapp.cs

```diff
@@ -41,6 +209,57 @@ private static void Scale(float a, Span<float> dst)
             }
         }
 
+        // dst = a * src
+        public static void Scale(float a, float[] src, float[] dst, int count)
+        {
+            Contracts.AssertNonEmpty(src);
```

Comment:
If AssertNonEmpty just checks for the length to be greater than 0 then this doesn't feel necessary to me, since you're asserting below that 0 < count and then that count <= src.Length, which automatically achieves that. If count is greater than 0 then it means src.Length must be greater than 0 as well.

Reply:
It is true that `0 < count && count <= src.Length` already takes care of checking if `src` is empty, and `Contracts.AssertNonEmpty<T>` doesn't seem to print any useful error messages here: https://github.com/dotnet/machinelearning/blob/d0664c1063539805c390a0f95231dcbfb87bc974/src/Microsoft.ML.Core/Utilities/Contracts.cs#L922-L927

It seems to me that we may want to first check whether `src` is non-empty.  If not, the program does not even bother to do the two checks for `count`.  `0 < count && count <= src.Length` sounds to me more like a restriction on `count` given the knowledge that `src` is a non-empty array.  If `src` were really empty, with the `AssertNonEmpty`, users could see from the stack trace that the line `Contracts.AssertNonEmpty(src)` is the problem, but without the `AsssertNonEmpty`, they would see from the stack trace that the line `Contracts.Assert(count <= src.Length)` is the problem, which might lead them to check for both `count` and `src`.

Reply:
Yes, let's check first wether `src` is non-empty and then just check if count <= src.Length.

```cs
Contracts.AssertNonEmpty(src);
Constracts.Assert(count <= src.Length);
```

As @tannergooding stated in  another comment, let's break out the && asserts.